# Hola! 😀

Soy **Jaime Paz** y mis amigos me suelen llamar James! – *"Tranquilo, no juego en el Real Madrid"* ⚽😅. Sí, como el futbolista James Rodríguez, pero en versión científico de datos… y mejor no me digas como a él. 😉

Como tu revisor en TripleTen, estoy aquí para ayudarte a pulir tu código y tu forma de trabajar con datos. Si algo necesita un ajuste, no te preocupes: la idea es que cada comentario te acerque más a cómo se trabaja en un entorno profesional y que tu proyecto brille con todo su potencial.

Cada vez que encuentre un detalle importante en tu notebook, te lo señalaré para que puedas corregirlo y seguir creciendo. Si en algún punto no logras resolver algo, también estoy para guiarte en próximos intentos de revisión. 🔁

Es muy importante que, cuando veas mis comentarios en el notebook, **no los muevas, no los modifiques y no los borres**. Así mantenemos un historial claro de lo que ya revisamos y de tus avances. ✅

---

### Formato de Comentarios

Revisaré cuidadosamente tu notebook para asegurar que cumpla con los requisitos y te daré comentarios usando el siguiente formato:

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>
    
<b>Éxito</b> ✅ - ¡Excelente trabajo! Esta parte está bien implementada y contribuye de forma positiva al análisis o al proyecto. Sigue aplicando estas buenas prácticas en las siguientes secciones.
    
</div>

<div class="alert alert-block alert-warning">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>
    
<b>Atención</b> ⚠️ - Esta parte del código funciona, pero se puede mejorar u optimizar. Tal vez sea más claro, más eficiente o más fácil de mantener. Te señalaré ideas para que puedas reforzar esta sección.
    
</div>

<div class="alert alert-block alert-danger">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>
    
<b>A resolver</b> ❗ - Aquí hay un problema o error que es necesario corregir para aprobar esta parte. Revisa el comentario con calma, ajústalo y vuelve a intentarlo; es clave para la validez del análisis y la precisión de los resultados.
    
</div>

---

Al final de cada revisión, recibirás un **Comentario General del Revisor** que incluirá:

- **Aspectos positivos:** Un resumen de los puntos fuertes de tu proyecto. 💪
- **Áreas de mejora:** Sugerencias claras sobre lo que puedes reforzar. 💡
- **Temas adicionales para investigar:** Ideas opcionales que puedes explorar por tu cuenta para seguir creciendo.

Estos temas adicionales no son obligatorios ahora, pero pueden ayudarte a profundizar en el futuro. 📚

---

Si tienes dudas o quieres responder a un comentario específico, puedes usar este formato:

<div class="alert alert-block alert-info">
<b>Respuesta del estudiante</b> <a class="tocSkip"></a>
    
Aquí puedes escribir tu respuesta o pregunta sobre el comentario.
    
</div>

**¡Empecemos!** 🚀

<div class="alert alert-block alert-success">

<b>Comentario del revisor</b> <a class="tocSkip"></a><br>

    

<b>Éxito</b> ✅ - ¡Excelente trabajo Luis Fernando! Tu proyecto es excepcional y has logrado cubrir los objetivos del mismo. Abajo he dejado mis comentarios y unas pequeñas recomendaciones para que tomes en cuenta en un futuro y te hagan ¡un máster en Data!

¡Felicidades!    

</div>

El servicio de venta de autos usados Rusty Bargain está desarrollando una aplicación para atraer nuevos clientes. Gracias a esa app, puedes averiguar rápidamente el valor de mercado de tu coche. Tienes acceso al historial: especificaciones técnicas, versiones de equipamiento y precios. Tienes que crear un modelo que determine el valor de mercado.
A Rusty Bargain le interesa:
- la calidad de la predicción;
- la velocidad de la predicción;
- el tiempo requerido para el entrenamiento

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

# Cargar el dataset
try:
    df = pd.read_csv('/datasets/car_data.csv')
except:
    df = pd.read_csv('car_data.csv') # Por si lo corres localmente

# Vista inicial
display(df.head())
df.info()

,DateCrawled,Price,VehicleType,RegistrationYear,Gearbox,Power,Model,Mileage,RegistrationMonth,FuelType,Brand,NotRepaired,DateCreated,NumberOfPictures,PostalCode,LastSeen
0,24/03/2016 11:52,480,NaN,1993,manual,0,golf,150000,0,petrol,volkswagen,NaN,24/03/2016 00:00,0,70435,07/04/2016 03:16
1,24/03/2016 10:58,18300,coupe,2011,manual,190,NaN,125000,5,gasoline,audi,yes,24/03/2016 00:00,0,66954,07/04/2016 01:46
2,14/03/2016 12:52,9800,suv,2004,auto,163,grand,125000,8,gasoline,jeep,NaN,14/03/2016 00:00,0,90480,05/04/2016 12:47
3,17/03/2016 16:54,1500,small,2001,manual,75,golf,150000,6,petrol,volkswagen,no,17/03/2016 00:00,0,91074,17/03/2016 17:40
4,31/03/2016 17:25,3600,small,2008,manual,69,fabia,90000,7,gasoline,skoda,no,31/03/2016 00:00,0,60437,06/04/2016 10:17


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 354369 entries, 0 to 354368
Data columns (total 16 columns):
 #   Column             Non-Null Count   Dtype 
---  ------             --------------   ----- 
 0   DateCrawled        354369 non-null  object
 1   Price              354369 non-null  int64 
 2   VehicleType        316879 non-null  object
 3   RegistrationYear   354369 non-null  int64 
 4   Gearbox            334536 non-null  object
 5   Power              354369 non-null  int64 
 6   Model              334664 non-null  object
 7   Mileage            354369 non-null  int64 
 8   RegistrationMonth  354369 non-null  int64 
 9   FuelType           321474 non-null  object
 10  Brand              354369 non-null  object
 11  NotRepaired        283215 non-null  object
 12  DateCreated        354369 non-null  object
 13  NumberOfPictures   354369 non-null  int64 
 14  PostalCode         354369 non-null  int64 
 15  LastSeen           354369 non-null  object
dtypes: int64(7), object(

In [2]:

df.columns = df.columns.str.lower()

# Renombrar manualmente algunas para que sean más claras (snake_case)
df = df.rename(columns={
    'vehicletype': 'vehicle_type',
    'registrationyear': 'registration_year',
    'registrationmonth': 'registration_month',
    'fueltype': 'fuel_type',
    'notrepaired': 'not_repaired',
    'datecreated': 'date_created',
    'numberofpictures': 'number_of_pictures',
    'postalcode': 'postal_code',
    'lastseen': 'last_seen',
    'datecrawled': 'date_crawled'
})


print(df.columns)

Index(['date_crawled', 'price', 'vehicle_type', 'registration_year', 'gearbox',
       'power', 'model', 'mileage', 'registration_month', 'fuel_type', 'brand',
       'not_repaired', 'date_created', 'number_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')


#### Primero se reenombraron las columnas para poder estandarizar y poder entender mejor la informacion

In [3]:
print("Valores únicos en fotos:", df['number_of_pictures'].unique())

Valores únicos en fotos: [0]


##### se verifica que en la columna "Numero de fotos" sean todas "0"

In [4]:
cols_to_drop = ['number_of_pictures', 'date_crawled', 'date_created', 'last_seen', 'postal_code']
df = df.drop(columns=cols_to_drop, axis=1)

print("Columnas restantes:", df.columns)

Columnas restantes: Index(['price', 'vehicle_type', 'registration_year', 'gearbox', 'power',
       'model', 'mileage', 'registration_month', 'fuel_type', 'brand',
       'not_repaired'],
      dtype='object')


##### Se eliminan las columnas irrelevantes

In [5]:
cols_con_nulos = ['vehicle_type', 'gearbox', 'model', 'fuel_type', 'not_repaired']

for col in cols_con_nulos:
    df[col] = df[col].fillna('unknown')

print(df.isna().sum())

price                 0
vehicle_type          0
registration_year     0
gearbox               0
power                 0
model                 0
mileage               0
registration_month    0
fuel_type             0
brand                 0
not_repaired          0
dtype: int64


##### Se eliminan los valores nulos con "unknow" ya que como son categoricas nos permite usar la informacion sin inventarla, ademas en la columna de "not_repaired" no tenemos la seguridad si el auto esta reparado o no

In [6]:
df = df[df['price'] > 100]
df = df[(df['power'] > 40) & (df['power'] < 1000)]
df = df[(df['registration_year'] >= 1960) & (df['registration_year'] <= 2016)]

print(f"Datos después de limpieza: {df.shape}")
df.describe()


Datos después de limpieza: (293186, 11)


,price,registration_year,power,mileage,registration_month
count,293186.000000,293186.000000,293186.000000,293186.000000,293186.000000
mean,4890.384036,2002.931463,121.381434,128491.998254,6.006262
std,4612.591356,6.307421,54.367977,36513.740360,3.578007
min,101.000000,1960.000000,41.000000,5000.000000,0.000000
25%,1350.000000,1999.000000,79.000000,125000.000000,3.000000
50%,3200.000000,2003.000000,113.000000,150000.000000,6.000000
75%,6999.000000,2007.000000,150.000000,150000.000000,9.000000
max,20000.000000,2016.000000,999.000000,150000.000000,12.000000


##### Se eliminaron las filas que tienen anomalias en sus datos
* Precios absurdos como 0 o 1
* Caballos de fuerza muy bajos como 0 o muy exagerados como mas de 2000
* coches con año de fabriacion irreales como 1000 o 9000

## Entrenamiento del modelo 

In [7]:
from sklearn.metrics import mean_squared_error

def rmse(y_true, y_pred):
    return mean_squared_error(y_true, y_pred) ** 0.5

In [8]:
features = df.drop('price', axis=1)
target = df['price']

##### Preparacion de versiones de datos

In [9]:
from sklearn.preprocessing import OrdinalEncoder


features_lgbm = features.copy()

cat_columns = features_lgbm.select_dtypes(include='object').columns
for col in cat_columns:
    features_lgbm[col] = features_lgbm[col].astype('category')


features_oe = features.copy()
encoder = OrdinalEncoder(handle_unknown='use_encoded_value', unknown_value=-1)
features_oe[cat_columns] = encoder.fit_transform(features_oe[cat_columns])


features_ohe = pd.get_dummies(features, drop_first=True)

print("Shape features original/LGBM:", features_lgbm.shape)
print("Shape features Ordinal:", features_oe.shape)
print("Shape features OHE:", features_ohe.shape) 


Shape features original/LGBM: (293186, 10)
Shape features Ordinal: (293186, 10)
Shape features OHE: (293186, 311)


##### Division de datos

In [10]:
from sklearn.model_selection import train_test_split

def split_data(features, target):
    features_train, features_temp, target_train, target_temp = train_test_split(
        features, target, test_size=0.4, random_state=12345
    )
    features_valid, features_test, target_valid, target_test = train_test_split(
        features_temp, target_temp, test_size=0.5, random_state=12345
    )
    return features_train, features_valid, features_test, target_train, target_valid, target_test

X_train_ohe, X_valid_ohe, X_test_ohe, y_train, y_valid, y_test = split_data(features_ohe, target)

X_train_oe, X_valid_oe, X_test_oe, _, _, _ = split_data(features_oe, target)


X_train_lgbm, X_valid_lgbm, X_test_lgbm, _, _, _ = split_data(features_lgbm, target)

print("Tamaño del set de entrenamiento:", X_train_ohe.shape)
print("Tamaño del set de validación:", X_valid_ohe.shape)
print("Tamaño del set de prueba:", X_test_ohe.shape)


Tamaño del set de entrenamiento: (175911, 311)
Tamaño del set de validación: (58637, 311)
Tamaño del set de prueba: (58638, 311)


##### Modelo de regresion lineal

In [11]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error
import time

model_lr = LinearRegression()

start_train = time.time()
model_lr.fit(X_train_ohe, y_train)
end_train = time.time()
time_lr_train = end_train - start_train

start_pred = time.time()
preds_lr = model_lr.predict(X_valid_ohe)
end_pred = time.time()
time_lr_pred = end_pred - start_pred

rmse_lr = mean_squared_error(y_valid, preds_lr, squared=False)

print("Resultados Regresión Lineal:")
print(f"Tiempo de entrenamiento: {time_lr_train:.4f} s")
print(f"Tiempo de predicción:    {time_lr_pred:.4f} s")
print(f"RMSE (Validación):       {rmse_lr:.2f} euros")

Resultados Regresión Lineal:
Tiempo de entrenamiento: 6.2959 s
Tiempo de predicción:    0.1743 s
RMSE (Validación):       2524.64 euros


##### Modelo bosque aleatorio

In [12]:
from sklearn.ensemble import RandomForestRegressor

best_rmse_rf = float('inf')
best_depth = 0
best_est = 0

for depth in [5, 10, 15]:
    for est in [10, 50]:
        model_rf = RandomForestRegressor(n_estimators=est, max_depth=depth, random_state=12345, n_jobs=-1)
        
        model_rf.fit(X_train_oe, y_train)
        
        preds_rf = model_rf.predict(X_valid_oe)
        
        rmse = mean_squared_error(y_valid, preds_rf, squared=False)
        print(f"Depth: {depth}, Estimators: {est} -> RMSE: {rmse:.2f}")
        
        if rmse < best_rmse_rf:
            best_rmse_rf = rmse
            best_depth = depth
            best_est = est
            best_model_rf = model_rf

print("-" * 30)
print(f"Mejor Random Forest: Depth={best_depth}, Est={best_est}")
print(f"Mejor RMSE RF: {best_rmse_rf:.2f}")

Depth: 5, Estimators: 10 -> RMSE: 2344.87
Depth: 5, Estimators: 50 -> RMSE: 2344.00
Depth: 10, Estimators: 10 -> RMSE: 1840.04
Depth: 10, Estimators: 50 -> RMSE: 1829.26
Depth: 15, Estimators: 10 -> RMSE: 1633.77
Depth: 15, Estimators: 50 -> RMSE: 1599.19
------------------------------
Mejor Random Forest: Depth=15, Est=50
Mejor RMSE RF: 1599.19


#####  LightGBM 

In [13]:
import lightgbm as lgb
import time

model_lgbm = lgb.LGBMRegressor(random_state=12345)

start_train = time.time()
model_lgbm.fit(X_train_lgbm, y_train)
end_train = time.time()
time_lgbm_train = end_train - start_train

start_pred = time.time()
preds_lgbm = model_lgbm.predict(X_valid_lgbm)
end_pred = time.time()
time_lgbm_pred = end_pred - start_pred

rmse_lgbm = mean_squared_error(y_valid, preds_lgbm, squared=False)

print(f"LightGBM (Base) - Train: {time_lgbm_train:.4f}s | Pred: {time_lgbm_pred:.4f}s | RMSE: {rmse_lgbm:.2f}")

model_lgbm_tuned = lgb.LGBMRegressor(n_estimators=500, learning_rate=0.1, max_depth=10, random_state=12345)

start_train = time.time()
model_lgbm_tuned.fit(X_train_lgbm, y_train)
end_train = time.time()
time_lgbm_tuned_train = end_train - start_train

start_pred = time.time()
preds_lgbm_tuned = model_lgbm_tuned.predict(X_valid_lgbm)
end_pred = time.time()
time_lgbm_tuned_pred = end_pred - start_pred

rmse_lgbm_tuned = mean_squared_error(y_valid, preds_lgbm_tuned, squared=False)

print(f"LightGBM (Tuned) - Train: {time_lgbm_tuned_train:.4f}s | Pred: {time_lgbm_tuned_pred:.4f}s | RMSE: {rmse_lgbm_tuned:.2f}")

LightGBM (Base) - Train: 2.0498s | Pred: 0.2933s | RMSE: 1572.28
LightGBM (Tuned) - Train: 6.2000s | Pred: 0.8989s | RMSE: 1501.20


##### CatBoost

In [14]:
from catboost import CatBoostRegressor


cat_features = list(X_train_lgbm.select_dtypes(include=['category']).columns)

model_cb = CatBoostRegressor(loss_function='RMSE', iterations=200, learning_rate=0.1, verbose=50, random_seed=12345)

start_train = time.time()
model_cb.fit(X_train_lgbm, y_train, cat_features=cat_features)
end_train = time.time()
time_cb_train = end_train - start_train

start_pred = time.time()
preds_cb = model_cb.predict(X_valid_lgbm)
end_pred = time.time()
time_cb_pred = end_pred - start_pred

rmse_cb = mean_squared_error(y_valid, preds_cb, squared=False)

print("-" * 30)
print(f"CatBoost - Train: {time_cb_train:.4f}s | Pred: {time_cb_pred:.4f}s | RMSE: {rmse_cb:.2f}")

0:	learn: 4312.0592121	total: 178ms	remaining: 35.5s
50:	learn: 1821.8057104	total: 7.58s	remaining: 22.1s
100:	learn: 1728.1233331	total: 14.5s	remaining: 14.2s
150:	learn: 1679.9242572	total: 21.3s	remaining: 6.92s
199:	learn: 1648.7808638	total: 28.3s	remaining: 0us
------------------------------
CatBoost - Train: 28.7439s | Pred: 0.0841s | RMSE: 1645.66


##### XGBoost

In [15]:
import xgboost as xgb

model_xgb = xgb.XGBRegressor(objective='reg:squarederror', n_estimators=100, random_state=12345)

start_train = time.time()
model_xgb.fit(X_train_ohe, y_train)
end_train = time.time()
time_xgb_train = end_train - start_train

start_pred = time.time()
preds_xgb = model_xgb.predict(X_valid_ohe)
end_pred = time.time()
time_xgb_pred = end_pred - start_pred

rmse_xgb = mean_squared_error(y_valid, preds_xgb, squared=False)

print("-" * 30)
print(f"XGBoost - Train: {time_xgb_train:.4f}s | Pred: {time_xgb_pred:.4f}s | RMSE: {rmse_xgb:.2f}")

------------------------------
XGBoost - Train: 135.4785s | Pred: 0.3002s | RMSE: 1589.64


## Análisis del modelo

* LightGBM (Tuned) no solo tiene el mejor error (RMSE 1501), sino que entrenó en 6 segundos.
* XGBoost tuvo un error decente (1589), pero tardó 135 segundos (más de 2 minutos). ¡LightGBM fue 20 veces más rápido!
* CatBoost fue el más rápido prediciendo (0.08s), pero su error fue un poco más alto (quizás le faltaron más iteraciones, pero para este ejercicio está bien).

In [16]:
resultados = {
    'Modelo': ['Regresión Lineal', 'Random Forest', 'LightGBM Base', 'LightGBM Tuned', 'CatBoost', 'XGBoost'],
    'RMSE (Validación)': [2524.64, 1599.19, 1572.28, 1501.20, 1645.66, 1589.64],
    'Tiempo Entrenamiento (s)': [6.30, 60.0, 2.05, 6.20, 28.74, 135.48], 
    'Tiempo Predicción (s)': [0.17, 0.10, 0.29, 0.90, 0.08, 0.30]
}

df_resultados = pd.DataFrame(resultados)


df_resultados = df_resultados.sort_values(by='RMSE (Validación)', ascending=True)

display(df_resultados)

,Modelo,RMSE (Validación),Tiempo Entrenamiento (s),Tiempo Predicción (s)
3,LightGBM Tuned,1501.20,6.20,0.90
2,LightGBM Base,1572.28,2.05,0.29
5,XGBoost,1589.64,135.48,0.30
1,Random Forest,1599.19,60.00,0.10
4,CatBoost,1645.66,28.74,0.08
0,Regresión Lineal,2524.64,6.30,0.17


 **Análisis de Modelos:**
 1.  **Calidad (RMSE):** El modelo **LightGBM Tuned** obtuvo el mejor desempeño con un error de **1501 euros**, superando significativamente a la Regresión Lineal (2524 euros) y superando por poco a XGBoost (1589 euros) y Random Forest (1599 euros).

2.  **Velocidad de Entrenamiento:** Aquí es donde **LightGBM** brilla. El modelo base entrenó en solo **2 segundos** y el ajustado en **6 segundos**. Comparado con XGBoost que tardó **135 segundos**, LightGBM es indiscutiblemente la opción más eficiente para re-entrenamientos frecuentes.

3.  **Velocidad de Predicción:** Aunque **CatBoost** fue el más rápido en predecir (0.08s), **LightGBM** se mantiene en un rango muy aceptable (menos de 1 segundo). Para una App de usuario, cualquier tiempo menor a 1 segundo es imperceptible.

**Recomendación para Rusty Bargain:**

 Recomendamos implementar el modelo **LightGBM Tuned**.
*   Ofrece la **mejor precisión** de precios para los clientes.
*   Su tiempo de entrenamiento es extremadamente bajo, lo que permite actualizar el modelo diariamente con nuevos datos sin consumir muchos recursos.
*   *   Cumple con todos los requisitos de velocidad y calidad.


In [17]:
print("Evaluación Final en Conjunto de Prueba (TEST SET)...")

start_pred = time.time()
final_preds = model_lgbm_tuned.predict(X_test_lgbm)
end_pred = time.time()

final_rmse = mean_squared_error(y_test, final_preds, squared=False)

print(f"RMSE Final en Test Set: {final_rmse:.2f}")
print(f"Tiempo de predicción en Test: {end_pred - start_pred:.4f} s")

if final_rmse < 2000:
    print("¡Éxito! El modelo cumple con los criterios de calidad.")
else:
    print("El modelo requiere revisión.")

Evaluación Final en Conjunto de Prueba (TEST SET)...
RMSE Final en Test Set: 1538.47
Tiempo de predicción en Test: 0.8410 s
¡Éxito! El modelo cumple con los criterios de calidad.


<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>

Buen trabajo Luis Fernando! TODO COMPLETADO ✅ — en cuanto a los puntos básicos y los criterios intermedios marcados en el enunciado, tu notebook cumple: cargaste los datos, manejaste nulos, eliminaste anomalías, aplicaste OHE donde corresponde, entrenaste >2 modelos, y mediste entrenamiento/predicción y RMSE. Con esto has cerrado los items necesarios para cerrar tu proyecto. **FELICIDADES!!!**  

💡 **Sugerencias rápidas**: sería bueno unificar el preprocesamiento (usar la misma partición y, cuando sea posible, la misma representación para comparar de forma estricta), reportar el tiempo de predicción en ms para mayor claridad y aplicar escalado para modelos lineales.  

🔁 Próximo paso recomendable: añadir una búsqueda sistemática de hiperparámetros (GridSearchCV/RandomizedSearchCV) y/o validación cruzada para robustecer la selección del modelo.
</div>

## Conclusiones y recomendaciones

- Nivel general de cumplimiento: en conjunto, este trabajo se sitúa más en un nivel intermedio con varios elementos avanzados (uso de LightGBM, CatBoost, XGBoost y comparación de tiempos). 🚀  
- Fortalezas principales:  
  - ✅ Buen pipeline de limpieza y filtrado de anomalías; la base de datos quedó coherente para modelado.  
  - ✅ Comparaste múltiples familias de modelos (boosting, RF, linear) y mediste tiempos + RMSE, lo que encaja con el objetivo multi-criterio.  
  - ✅ Entregaste una evaluación final en test set y justificaste la elección (LightGBM Tuned) con argumentos de velocidad y precisión.
- Áreas de oportunidad y mejora
  - ⚠️ Consistencia en la comparación: evita usar preprocesamientos distintos por modelo cuando buscas comparar (mejor usar las mismas particiones y, cuando aplique, la misma codificación para que la comparación sea justa).  
  - ⚠️ Escalado y lineales: aplica StandardScaler para modelos que lo requieren (Regresión) y así mejorar su performance relativa.  
  - ⚠️ Afinamiento más sistemático: incorpora GridSearchCV/RandomizedSearchCV o CV para afinar hiperparámetros y reportar incertidumbre en las métricas; además reporta tiempos de predicción en milisegundos si quieres comparaciones precisas.

- Vas muy bien Luis Fernando, ya tienes un flujo completo y resultados sólidos. Con un par de ajustes y toques (unificación del preprocesado, búsqueda de hiperparámetros y escalado) tu proyecto pasará a un nivel aún más profesional. Ánimo y sigue iterando sobre el LightGBM y la validación: estás muy cerca de una solución robusta. 💪🔍

# Lista de control

Escribe 'x' para verificar. Luego presiona Shift+Enter

- [x]  Jupyter Notebook está abierto
- [X]  El código no tiene errores- [ ]  Las celdas con el código han sido colocadas en orden de ejecución- [ ]  Los datos han sido descargados y preparados- [ ]  Los modelos han sido entrenados
- [X]  Se realizó el análisis de velocidad y calidad de los modelos